kafkaesk/retry.py

```diff
@@ -0,0 +1,354 @@
+
+    * RETRY_HANDLER_FORWARD - Note: This is implemented by RetryHandler._forward_message
+    * RETRY_HANDLER_DROP - Note: this is implemented by RetryHandler._drop_message
+    * RETRY_CONSUMER_TOPIC_OFFSET
+    * RETRY_CONSUMER_MESSAGE_LEAD_TIME - Note: If a RetryHandler's consumer expects a delay,
+        this delay should be subtracted from the lead time
+    * RETRY_CONSUMER_CONSUMED_MESSAGE_TIME
+    * RETRY_CONSUMER_CONSUMED_MESSAGES
+
+    See `kafkaesk.metrics` for more information on each of these metrics
+    """
+
+    logger = logging.getLogger("kafkaesk.retry.retry_handler")
+
+    def __init__(self) -> None:
+        self._ready = False
+
+    async def initialize(self, policy: RetryPolicy) -> None:
+        self.logger.debug(f"{self.__class__.__name__} retry handler initialized")
+        self._ready = True
+
+    async def finalize(self) -> None:
+        self.logger.debug(f"{self.__class__.__name__} retry handler finalized")
+        self._ready = False
+
+    async def __call__(
+        self,
+        policy: RetryPolicy,
+        handler_key: str,
+        retry_history: RetryHistory,
+        record: ConsumerRecord,
+        exception: Exception,
+    ) -> None:
+        raise NotImplementedError
+
+    async def _raise_message(
+        self,
+        policy: RetryPolicy,
+        retry_history: RetryHistory,
+        record: ConsumerRecord,
+        exception: Exception,
+    ) -> None:
+        self.logger.critical(
+            f"{self.__class__.__name__} handler recieved exception, re-raising exception {record}"
+        )
+        RETRY_HANDLER_RAISE.labels(
+            stream_id=record.topic,
+            partition=record.partition,
+            group_id=policy.subscription.group,
+            handler=self.__class__.__name__,
+            exception=exception.__class__.__name__,
+        ).inc()
+        raise exception from exception
+
+    async def _drop_message(
+        self,
+        policy: RetryPolicy,
+        retry_history: RetryHistory,
+        record: ConsumerRecord,
+        exception: Exception,
+    ) -> None:
+        self.logger.warn(
+            f"{self.__class__.__name__} handler recieved exception, dropping message {record}",
+            exc_info=exception,
+        )
+        RETRY_HANDLER_DROP.labels(
+            stream_id=record.topic,
+            partition=record.partition,
+            group_id=policy.subscription.group,
+            handler=self.__class__.__name__,
+            exception=exception.__class__.__name__,
+        ).inc()
+
+    async def _forward_message(
+        self,
+        policy: RetryPolicy,
+        retry_history: RetryHistory,
+        record: ConsumerRecord,
+        exception: Exception,
+        forward_stream_id: str,
+    ) -> None:
+        self.logger.info(
+            f"{self.__class__.__name__} handler forwarding message to {forward_stream_id}: {record}"
+        )
+        await policy.app.publish(
```

Comment:
What do you think about allowing string formatters here?

for example:

```
forward_stream_id.format(record=record, group_id=group_id)
```

When the `forward_stream_id` might be something like `{record.topic}-{group_id}.forward`

Because remember, these subscriptions might be against many topics.

I'm okay if we want to leave it the way it is for now though. Maybe when you do retries, it is something where you want to be explicit on topic names to forward messages to because it's a particular topic + group id combo that caused the error.

Reply:
This code is in the base `RetryHandler` class and available as a utility method.  The `forward_stream_id` is passed in by the sub-class and can be whatever the implementation needs it to be.

The logging line starting on 297 could probably have its formatting improved to not cause confusion as it looks like `record` is being added to `forward_stream_id` when it's just providing additional context within the logging call.

Reply:
I'm just suggesting that we might want a way to define dynamic streams based on the record it's handling.

You saying this can be accomplished by sub-classes and providing your own there? Okay, that's fine for me.

It's something we can iterate on nonetheless.